### PR TITLE
Improved index fetching, supporting multiple versions

### PR DESCRIPTION
**Pull request by stearm**
This PR adds a method to retrieve only the Lyra index corresponding to the selected documentation version, as discussed in https://github.com/LyraSearch/plugin-docusaurus/issues/77#issuecomment-1380288513. For each documentation version, we create an index file.

Next steps:
- add the 'pattern' option to choose which indexes files to generate, and figure out how to add tests.
- once Lyra is updated to v0.4.x, support it

This is a step closer to #77.

### CODE REVIEW

File: src/client/theme/SearchBar/utils.ts [hunk 1]
```diff
@@ -0,0 +1,4 @@
+export function getSelectedDocsVersion() {
+  const url = window.location.href.toString()
+  return url.match(/\/(v[^/]+)/)?.[1] || 'latest'
```
Comment (Eomm) [inline src/client/theme/SearchBar/utils.ts:3]: Note that the version string is a customizable thing in docusaurus

- Jest sets it as `next` or `29.x` https://jestjs.io/
- Docusaurus sets it as `2.1.0` https://docusaurus.io/docs/2.1.0

I think a safer getter should get the string after `docs/<version>/` 

Even difficult: the latest version, by default, does not have the version as URL path:
`https://docusaurus.io/docs/configuration`

The `latest` thing is a fastify website's legacy

File: src/server/index.ts [hunk 2]
```diff
@@ -9,6 +10,34 @@ import { ResolveSchema } from '@lyrasearch/lyra/dist/esm/src/types'
...
+      .sync(`${outDir}/**/*/`)
+      .map(folder => folder.slice(folder.indexOf(outDir) + outDir.length))
+    const versionFolders = folders.filter(folder =>
+      semanticVersionRegex.test(basename(folder))
+    )
+
+    if (glob.sync(`${outDir}/docs`).length) {
+      return versionFolders.map(folder => [folder])
+    } else {
+      return [
```
Comment (Ceres6) [inline src/server/index.ts:26]: I think this is a bit hard to follow. There should be a simpler way of getting folders of versions and the rest. In case there's not extracting it to functions which names states the purpose could make it easier to read.

File: src/client/theme/SearchBar/utils.ts (hunk 1 above)
Comment (Ceres6) [reply src/client/theme/SearchBar/utils.ts:3]: Probably there is a way to get this from config (the latest/next thing, I mean). About the semantic versioning I think the previous regex that you had, adding optional "v" covered the option of `29.x`

**Review by Ceres6 — COMMENTED** [review]
As it is now it only indexes docs. Before it also indexed blog and regular pages. We should probably index those in a separate index and have it as a second source independent of the version we're at.

I'm not sure if this can happen in sites without docs folder, but in that case there should be some way to tell apart doc folders from regular/blog ones

File: src/client/theme/SearchBar/utils.ts (hunk 1 above)
Comment (Eomm) [reply src/client/theme/SearchBar/utils.ts:3]: I think it would work 👍🏽 

> if there's vX.X.X or X.X.X it's a "normal" version

Just to mention the `29.x` jest's version

**Comment by stearm** [comment]
@Ceres6 I refactored the 'getDocVersionsFolders' function, and I believe it is now more readable. @Eomm, we now support versions without a 'v' as well as those used by Jest.

I'll start working on the blog and pages during next days.

**Comment by stearm** [comment]
The last missing part is multi-locales support; the idea is to create indexes for each language. What I'm not sure about is how to understand which language the user selects, to retrieve the correct index. Reading the DOM is not a good solution imho. Any suggestion is appreciated.

**Comment by Eomm** [comment]
You can read the `<html lang="fr"` attribute - it is set by docusaurus to let you know which lang is loaded

File: src/server/index.ts [hunk 3]
```diff
@@ -7,9 +8,87 @@ import { create } from '@lyrasearch/lyra'
...
+
+    const latest = folders.filter(
+      folder =>
+        !versionFoldersRegExp.test(folder) && folder.indexOf(blogPath) < 0
+    )
+
+    if (glob.sync(`${outDir}/docs`).length) {
+      return {
+        versionFolders,
+        latest: latest.filter(
```
Comment (Ceres6) [inline src/server/index.ts:38]: If docs exists then you don't need to do the first filter, you can directly take everything inside it, am I wrong?

File: src/server/index.ts [hunk 4]
```diff
@@ -7,9 +8,87 @@ import { create } from '@lyrasearch/lyra'
 import { SectionSchema } from '../types'
 import { ResolveSchema } from '@lyrasearch/lyra/dist/esm/src/types'
 import { defaultHtmlSchema } from '@lyrasearch/plugin-parsedoc'
-import { INDEX_FILE, PLUGIN_NAME } from '../shared'
+import { INDEX_FILE, PLUGIN_NAME, versionFoldersRegExp } from '../shared'
+
+const getDocVersionsFolders = ({
```
Comment (Ceres6) [inline src/server/index.ts:13]: I think this function does too many things and now we have fresh knowledge of all the possible structures but in the future, or anyone who is not us will probably have a hard time understanding what's going on. I would extract the parts of the function into other functions with explanatory names so each step is easier to follow.

File: src/server/index.ts [hunk 5]
```diff
@@ -7,9 +8,87 @@ import { create } from '@lyrasearch/lyra'
...
+    }
+  } catch (err) {
+    console.error('Reading versions folders failed, cannot generate indexes.')
+    return { versionFolders: [] }
+  }
+}
+
+async function retrieveBlogIndex(outDir: string, blogPath: string) {
+  return retrieveIndex(
+    [`${`${outDir}/${blogPath}`}/**/*.html`],
```
Comment (Ceres6) [inline src/server/index.ts:57]: Why nesting template strings?
```ts
[`${outDir}/${blogPath}/**/*.html`]

File: src/server/index.ts [hunk 6]
```diff
@@ -29,26 +108,53 @@ const docusaurusLyraSearchPlugin = (
     const index = await retrieveDevIndex(allContent)
     await actions.setGlobalData(index)
   },
-  async postBuild({ outDir }) {
-    const index = await retrieveIndex('**/*.html', pageRouteFactory)
-    return writeIndex(outDir, index)
+  async postBuild({ outDir, plugins }) {
+    const blogPath = (plugins.find(
+      plugin => plugin.name === 'docusaurus-plugin-content-blog'
+    )?.options.routeBasePath || '') as string
```
Comment (Ceres6) [inline src/server/index.ts:114]: as per [docusaurus blog documentation](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-blog) the default path is `blog` (maybe it always have a value and we are just testing for the existence of the plugin).
Also here we are looking for file path so we should probably be using `path` instead of `routeBasePath` which is used for urls (maybe I'm wrong and docusaurus in the build transforms the `path` directory into the `routeBasePath` directory).

Probably we should follow the same approach with docs.

**Review by Ceres6 — CHANGES_REQUESTED** [review]

**Comment by frsgrn** [comment]
Supporting search with multi-instance docs (using multiple docs plugins) would be great, really hoping that this gets merged.
